File: modules/domain_recon.py
# -*- coding: utf-8 -*-

# MODULES AND/OR LIBRARIES
import socket
import dns.resolver
import whois
import time
from modules.logging_config import debug, error, info

##############################

# IP ADDRESS

##############################

def get_ip_addresses(target_domain_name):
    try:
        ip_addresses = []
        for result in socket.getaddrinfo(target_domain_name, 80):
            ip_addresses.append(result[4][0])
        debug(f"Raw IP addresses: {ip_addresses}")
        ip_addresses = list(set(ip_addresses))
        return ip_addresses
    except Exception as e:
        error(f"Error: {e}")

##############################

# MX RECORDS

##############################

def get_mx_records(target_domain_name):
    try:
        mx_records_raw = dns.resolver.resolve(target_domain_name, 'MX')
        mx_records = []
        for raw_data in mx_records_raw:
            mx_records.append(str(raw_data.exchange))
        debug(f"Raw mx records: {mx_records}")
        return mx_records
    except Exception as e:
        error(f"Error: {e}")

##############################

# NAMESERVERS

##############################

def get_nameservers(target_domain_name):
    try:
        nameservers_raw = dns.resolver.resolve(target_domain_name, 'NS')
        nameservers = []
        for raw_data in nameservers_raw:
            nameservers.append(str(raw_data.target))
        debug(f"Raw nameservers data: {nameservers}")
        return nameservers
    except Exception as e:
        error(f"Error: {e}")

##############################

# WHOIS

##############################

def get_whois_info(target_domain_name):
    try:
        whois_raw = whois.whois(target_domain_name)
        debug(f"Raw WHOIS data: {whois_raw}")
        whois_data = {
            "Registrar": whois_raw.registrar if whois_raw.name else 'N/A',
            "Registrant": whois_raw.name if whois_raw.name else 'N/A',
            "Creation Date": whois_raw.creation_date if whois_raw.creation_date else 'N/A',
            "Expiration Date": whois_raw.expiration_date if whois_raw.expiration_date else 'N/A',
            "Nameservers": whois_raw.name_servers if whois_raw.name_servers else [],
            "Emails": whois_raw.emails if whois_raw.emails else [],
            "Registrant Organization": whois_raw.org or 'N/A',
            "Abuse Contact": whois_raw.get('abuse_contact_email') or 'N/A',
            "DNSSEC": whois_raw.dnssec if whois_raw.dnssec else 'Not implemented',
            "Domain Status": whois_raw.status if whois_raw.status else 'N/A',
        }
        return whois_data
    except Exception as e:
        error(f"WHOIS lookup error: {e}")

##############################

# COLLECT DOMAIN INFO

##############################

def collect_domain_info(target_domain_name):
    start_time = time.time()
    print(f"[+] Collecting domain info for: {target_domain_name}")
    try:
        domain_info = {
            "IP Addresses": get_ip_addresses(target_domain_name=target_domain_name),
            "MX Records": get_mx_records(target_domain_name=target_domain_name),
            "Nameservers": get_nameservers(target_domain_name=target_domain_name),
            "Whois": get_whois_info(target_domain_name=target_domain_name),
        }
    except KeyboardInterrupt:
        print("\n[!] Enumeration interrupted. Returning partial results.\n")
    finally:
        duration = time.time() - start_time
        print(f"\n[✓] Domain information collection complete. Duration: {duration:.2f} seconds.\n")
        return domain_info

File: modules/domain_recon.go
package modules

import (
	"fmt"
	"log"
	"net"
	"os"
	"regexp"
	"strings"
	"time"

	"github.com/likexian/whois"
	"github.com/likexian/whois-parser"
)

// ##############################
//
// # HELPER FUNCTIONS
//
// ##############################

func cleanDate(dateStr string) string {
	if dateStr == "" {
		return "N/A"
	}
	parsedTime, err := time.Parse(time.RFC3339, dateStr)
	if err != nil {
		return dateStr
	}
	return parsedTime.Format("2006-01-02 15:04:05")
}

func fileExists(filename string) bool {
	info, err := os.Stat(filename)
	if os.IsNotExist(err) {
		return false
	}
	return !info.IsDir()
}

func extractEmails(raw string) []string {
	emailRegex := regexp.MustCompile(`[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}`)
	matches := emailRegex.FindAllString(raw, -1)
	unique := make(map[string]bool)
	for _, email := range matches {
		unique[email] = true
	}
	var emails []string
	for email := range unique {
		emails = append(emails, email)
	}
	return emails
}

func extractField(raw, fieldName string) string {
	lines := strings.Split(raw, "\n")
	fieldName = strings.ToLower(fieldName)
	for _, line := range lines {
		lowerLine := strings.ToLower(line)
		if strings.Contains(lowerLine, fieldName) {
			parts := strings.SplitN(line, ":", 2)
			if len(parts) == 2 {
				return strings.TrimSpace(parts[1])
			}
		}
	}
	return ""
}

// ##############################
//
// # IP ADDRESS
//
// ##############################

func GetIPAddresses(targetDomainName string) {
	fmt.Println("=== IP Addresses ===")
	addresses, err := net.LookupIP(targetDomainName)
	if err != nil {
		log.Println(err)
		return
	}
	for index, address := range addresses {
		fmt.Printf("%d) %s\n", index+1, address)
	}
}

// ##############################
//
// # NAMESERVERS
//
// ##############################

func GetNameservers(targetDomainName string) {
	fmt.Println("=== Nameservers ===")
	nsRecords, err := net.LookupNS(targetDomainName)
	if err != nil {
		log.Println(err)
		return
	}
	for index, nsRecord := range nsRecords {
		fmt.Printf("%d) %s\n", index+1, nsRecord.Host)
	}
}

// ##############################
//
// # MX RECORDS
//
// ##############################

func GetMXRecords(targetDomainName string) {
	fmt.Println("=== MX Records ===")
	mxRecords, err := net.LookupMX(targetDomainName)
	if err != nil {
		log.Println(err)
		return
	}
	for index, mxRecord := range mxRecords {
		fmt.Printf("%d) %s priority: %d\n", index+1, mxRecord.Host, mxRecord.Pref)
	}
}

// ##############################
//
// # WHOIS
//
// ##############################

func GetWHOIS(targetDomainName string) {
	fmt.Println("=== WHOIS ===")
	rawResult, err := whois.Whois(targetDomainName)
	if err != nil {
		log.Println("WHOIS error:", err)
		return
	}

	parsed, err := whoisparser.Parse(rawResult)
	if err != nil {
		log.Println("Parse error:", err)
		return
	}

	fmt.Println("Registrar:", parsed.Registrar.Name)

	registrant := parsed.Registrant.Name
	if registrant == "" {
		registrant = "None"
	}
	fmt.Println("Registrant:", registrant)

	creationDate := cleanDate(parsed.Domain.CreatedDate)
	expirationDate := cleanDate(parsed.Domain.ExpirationDate)
	fmt.Println("Creation Date:", creationDate)
	fmt.Println("Expiration Date:", expirationDate)

	// Nameservers (extract manually from raw WHOIS)
	nsRaw := extractField(rawResult, "Name Server")
	if nsRaw == "" {
		nsRaw = extractField(rawResult, "Name Servers")
	}
	if nsRaw != "" {
		fmt.Println("Nameservers:", nsRaw)
	} else {
		fmt.Println("Nameservers: N/A")
	}

	// Emails
	emails := extractEmails(rawResult)
	if len(emails) > 0 {
		fmt.Printf("Contact Emails: %s\n", strings.Join(emails, ", "))
	} else {
		fmt.Println("Contact Emails: N/A")
	}

	// Registrant organization
	org := parsed.Registrant.Organization
	if org == "" {
		org = extractField(rawResult, "Registrant Organization")
	}
	if org == "" {
		org = "N/A"
	}
	fmt.Println("Registrant Organization:", org)

	// Abuse contact
	abuseContact := extractField(rawResult, "Abuse Contact Email")
	if abuseContact == "" {
		abuseContact = "N/A"
	}
	fmt.Println("Abuse Contact:", abuseContact)

	// DNSSEC
	dnssec := extractField(rawResult, "DNSSEC")
	if dnssec == "" {
		dnssec = "Not implemented"
	}
	fmt.Println("DNSSEC:", dnssec)

	// Domain status
	status := extractField(rawResult, "Status")
	if status == "" {
		status = "N/A"
	}
	fmt.Println("Domain Status:", status)
}

// ##############################
//
// # FULL DOMAIN INFO
//
// ##############################

func GetDomainInfo(targetDomainName string) {
	startTime := time.Now()

	GetIPAddresses(targetDomainName)
	GetNameservers(targetDomainName)
	GetMXRecords(targetDomainName)
	GetWHOIS(targetDomainName)

	duration := time.Since(startTime).Seconds()
	fmt.Printf("\nDuration: %.2f seconds.\n", duration)
}